### ISSUE TITLE
Return Value of verifyconnect Should Not Be Ignored

### ISSUE THREAD

**Issue by churchill-zhang** [curl/curl#15084]
### I did this

```
 rc = SOCKET_WRITABLE(ctx->sock, 0);

  if(rc == 0) { /* no connection yet */
    CURL_TRC_CF(data, cf, "not connected yet");
    return CURLE_OK;
  }
  else if(rc == CURL_CSELECT_OUT || cf->conn->bits.tcp_fastopen) {
    if(verifyconnect(ctx->sock, &ctx->error)) {
      /* we are connected with TCP, awesome! */
      ctx->connected_at = Curl_now();
      set_local_ip(cf, data);
      *done = TRUE;
      cf->connected = TRUE;
      CURL_TRC_CF(data, cf, "connected");
      return CURLE_OK;
    }
  }
  else if(rc & CURL_CSELECT_ERR) {
    (void)verifyconnect(ctx->sock, &ctx->error);
    result = CURLE_COULDNT_CONNECT;
  }
```

### I expected the following

```
/* check socket for connect */
  rc = SOCKET_WRITABLE(ctx->sock, 0);

  if(rc == 0) { /* no connection yet */
    CURL_TRC_CF(data, cf, "not connected yet");
    return CURLE_OK;
  }
  else if((rc & CURL_CSELECT_ERR) || rc == CURL_CSELECT_OUT || cf->conn->bits.tcp_fastopen) {
    if(verifyconnect(ctx->sock, &ctx->error)) {
      /* we are connected with TCP, awesome! */
      ctx->connected_at = Curl_now();
      set_local_ip(cf, data);
      *done = TRUE;
      cf->connected = TRUE;
      CURL_TRC_CF(data, cf, "connected");
      return CURLE_OK;
    }
    if(rc & CURL_CSELECT_ERR) {
      result = CURLE_COULDNT_CONNECT;
    }
  }
```
Based on the definition and implementation of the verifyconnect function, if the function returns true, it indicates that the socket is currently usable. Previously, there was an issue with SOCKET_WRITABLE, but it can now be caught and has been restored. Therefore, I believe the return value of verifyconnect should not be ignored, and CURLE_COULDNT_CONNECT should not be returned. 

### curl/libcurl version

master

### operating system

Android and iOS

**Comment by bagder** [comment]
First: your change does not affect the return value of verifyconnect(), which makes the title odd and misleading. 

Then: should the fix not to be to rather make sure we care for the `CURL_CSELECT_OUT` bit and not the value?

That is, is the fix not more properly done like this?
~~~diff
--- a/lib/cf-socket.c
+++ b/lib/cf-socket.c
@@ -1347,11 +1347,11 @@ static CURLcode cf_tcp_connect(struct Curl_cfilter *cf,
 
   if(rc == 0) { /* no connection yet */
     CURL_TRC_CF(data, cf, "not connected yet");
     return CURLE_OK;
   }
-  else if(rc == CURL_CSELECT_OUT || cf->conn->bits.tcp_fastopen) {
+  else if((rc & CURL_CSELECT_OUT) || cf->conn->bits.tcp_fastopen) {
     if(verifyconnect(ctx->sock, &ctx->error)) {
       /* we are connected with TCP, awesome! */
       ctx->connected_at = Curl_now();
       set_local_ip(cf, data);
       *done = TRUE;

**Comment by churchill-zhang** [comment]
> First: your change does not affect the return value of verifyconnect(), which makes the title odd and misleading.
> 
> Then: should the fix not to be to rather make sure we care for the `CURL_CSELECT_OUT` bit and not the value?
> 
> That is, is the fix not more properly done like this?
> 
> ```diff
> --- a/lib/cf-socket.c
> +++ b/lib/cf-socket.c
> @@ -1347,11 +1347,11 @@ static CURLcode cf_tcp_connect(struct Curl_cfilter *cf,
>  
>    if(rc == 0) { /* no connection yet */
>      CURL_TRC_CF(data, cf, "not connected yet");
>      return CURLE_OK;
>    }
> -  else if(rc == CURL_CSELECT_OUT || cf->conn->bits.tcp_fastopen) {
> +  else if((rc & CURL_CSELECT_OUT) || cf->conn->bits.tcp_fastopen) {
>      if(verifyconnect(ctx->sock, &ctx->error)) {
>        /* we are connected with TCP, awesome! */
>        ctx->connected_at = Curl_now();
>        set_local_ip(cf, data);
>        *done = TRUE;
> ```
#15080 
Setting -DHAVE_POLL_H=ON -DHAVE_POLL_FINE=ON only resolves the issue on the Android platform. The problem persists on iOS. The cf-socket uses `rc = SOCKET_WRITABLE(ctx->sock, 0);` to check if the socket is working correctly. On iOS, this only follows the select branch, and VERIFY_SOCK will inevitably set errno and return -1 for socket file descriptors above 1024. In my project, which uses the https://curl.se/libcurl/c/multi-uv.html multi mode, socket file descriptors larger than 1024 work correctly, so verifyconnect returns TRUE. Therefore, it should not inform the caller of a CURLE_COULDNT_CONNECT error, but should instead correctly return CURLE_OK.

**Comment by churchill-zhang** [comment]
### Steps to Reproduce:

1. Modify the application's maximum file descriptor limit.
```
struct rlimit rlim;
if (getrlimit(RLIMIT_NOFILE, &rlim) == 0) {
  std::cout << "Soft limit: " << rlim.rlim_cur << std::endl;
  std::cout << "Hard limit: " << rlim.rlim_max << std::endl;
} else {
  std::cout << "Unable to get file descriptor limits" << std::endl;
}
rlim.rlim_cur = 32768;
rlim.rlim_max = 32768;
if (setrlimit(RLIMIT_NOFILE, &rlim) == -1) {
  std::cout << "Unable to set file descriptor limits" << std::endl;
}
```
2. Consume 1024 system file descriptors.
```
std::vector<int> fds;
int count = 1024;
while (count >= 0) {
  int fd = dup(STDOUT_FILENO);
  if (fd == -1) {
     std::cerr << "Cannot create new file descriptor" << std::endl;
     break;
  }
  fds.push_back(fd);
  count -= 1;
  std::cout << "Created file descriptor: " << fd << std::endl;
}
```
3. To construct a network request using a methodology similar to the one illustrated on [this page](https://curl.se/libcurl/c/multi-uv.html)

**Comment by bagder** [comment]
The description of this bug is wrong. The point was (I believe) that because the iOS build used select(), the `SOCKET_WRITABLE` macro would return an error and then the connection verification does not work.

This is now fixed via #15096 (c72cefea0fadaf4114a0036c86005ee5739ec30a) so that Apple operating systems also get to use poll() and thus can deal with large amount of file descriptors better.
